llvm/lib/CodeGen/GlobalISel/Utils.cpp

```diff
@@ -1724,6 +1724,39 @@ bool llvm::isPreISelGenericFloatingPointOpcode(unsigned Opc) {
   }
 }
 
+/// Shifts return poison if shiftwidth is larger than the bitwidth.
+static bool shiftAmountKnownInRange(Register ShiftAmount,
+                                    const MachineRegisterInfo &MRI) {
+  LLT Ty = MRI.getType(ShiftAmount);
+
+  if (Ty.isScalableVector())
+    return false; // Can't tell, just return false to be safe
+
+  if (Ty.isScalar()) {
+    std::optional<ValueAndVReg> Val =
+        getIConstantVRegValWithLookThrough(ShiftAmount, MRI);
+    if (!Val)
+      return false;
+    return Val->Value.ult(Ty.getScalarSizeInBits());
+  }
+
+  GBuildVector *BV = getOpcodeDef<GBuildVector>(ShiftAmount, MRI);
+  if (!BV)
+    return false;
+
+  unsigned Sources = BV->getNumSources();
```

Comment:
Should add a comment explaining why `GBuildVector` specifically is being checked.

Reply:
It is the source of a constant vector. If the shift type is a vector, we need to check the vector shift amount aka G_BUILD_VECTOR.

Reply:
See https://github.com/llvm/llvm-project/blob/d776346afe790e5d51ca6c6e2238a6ba91d130a1/llvm/lib/Analysis/ValueTracking.cpp#L7010
and
https://github.com/llvm/llvm-project/blob/53b904823cba4e93d58793b0804407f812477724/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp#L3073